### PR TITLE
Some new features are expected to come in handy.

### PR DESCRIPTION
**Pull request by oblivioned**
I found some problems in the process of using, and I fixed them according to my idea.

Also, a class called "ContractInvocation" has been added, which can perform some convenient operations for deployed smart contracts, such as such an example.

```dart
final invoker = erc20.transfer(
          toAddress,
          EtherAmount.fromEther(1).getInWei
      );

print('  > TryCallReceipt: ${await invoker.call(from: fromAddress)}');
print('  > EstimateGas: ${await invoker.estimateGas(from: fromAddress)}');
....
```

Of course, it can also be used in this way.

```dart
await erc20.transfer( toAddress,  EtherAmount.fromEther(1).getInWei ).call();

await erc20.transfer( toAddress,  EtherAmount.fromEther(1).getInWei ).estimateGas();

/// or sent it
await erc20.transfer( toAddress,  EtherAmount.fromEther(1).getInWei ).send(....);
```
For more details, see the code in "test/core/deployed_contract_test.dart".

### CODE REVIEW

File: lib/src/contracts/abi/abi.dart [hunk 1]
```diff
@@ -218,18 +218,34 @@ class ContractFunction {
   /// * uint<x> and int<x> will accept a dart int
   ///
   /// Other types are not supported at the moment.
-  Uint8List encodeCall(List<dynamic> params) {
-    if (params.length != parameters.length) {
-      throw ArgumentError.value(
-          params.length, 'params', 'Must match function parameters');
-    }
+  Uint8List encodeCall([List<dynamic> params]) {
```
Comment (simolus3) [inline lib/src/contracts/abi/abi.dart:221]: I don't think `params` should ever be null. If you want to make it optional, let's add a default parameter:

```suggestion
  Uint8List encodeCall([List<dynamic> params = const []]) {
```

File: lib/src/contracts/abi/abi.dart [hunk 2]
```diff
@@ -218,18 +218,34 @@ class ContractFunction {
...
 
     final sink = LengthTrackingByteSink()
-      //First four bytes to identify the function with its parameters
-      ..add(keccakUtf8(encodeName()).sublist(0, 4));
+    //First four bytes to identify the function with its parameters
+    ..add(keccakUtf8(encodeName()).sublist(0, 4));
+
+    if ( parameters == null || parameters.isEmpty ) {
```
Comment (simolus3) [inline lib/src/contracts/abi/abi.dart:227]: See above - the `parameters == null` and `params == null` check should be removed. Is there a reason to have this special check either way? If you want to optimize for the "empty parameters" case, I would probably write it like this:
```dart
//First four bytes to identify the function with its parameters
final functionIdentifier = keccakUtf8(encodeName()).sublist(0, 4));

if (params.isEmpty) {
  // Skip encoding arguments when we now there aren't any
  assert(parameters.isEmpty);
  return functionIdentifier;
} else {
  final link = LengthTrackingByteSink()..add(functionIdentifier);
  TupleType(parameters.map((param) => param.type).toList()).encode(params, sink);
  return sink.asBytes();
}
```

File: lib/src/contracts/contract_invoker.dart [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+    final String reason;
+
+    const ExternalABI([this.reason = '']);
+}
+
+const readonly = ExternalABI('external pure or view function');
+
+const readwrite = ExternalABI('external function');
+
+const readwritePayable = ExternalABI('external payable function');
```
Comment (simolus3) [inline lib/src/contracts/contract_invoker.dart:14]: I think we already have the `StateMutability` enum for this.

File: lib/src/core/amount.dart [hunk 4]
```diff
@@ -62,6 +62,14 @@ class EtherAmount {
     return EtherAmount.inWei(parsedAmount * _factors[unit]);
   }
 
+  factory EtherAmount.fromEther(dynamic amount) => EtherAmount.fromUnitAndValue(EtherUnit.ether, amount);
+  factory EtherAmount.fromFinney(dynamic amount) => EtherAmount.fromUnitAndValue(EtherUnit.finney, amount);
+  factory EtherAmount.fromSzabo(dynamic amount) => EtherAmount.fromUnitAndValue(EtherUnit.szabo, amount);
+  factory EtherAmount.fromGwei(dynamic amount) => EtherAmount.fromUnitAndValue(EtherUnit.gwei, amount);
+  factory EtherAmount.fromMwei(dynamic amount) => EtherAmount.fromUnitAndValue(EtherUnit.mwei, amount);
+  factory EtherAmount.fromKwei(dynamic amount) => EtherAmount.fromUnitAndValue(EtherUnit.kwei, amount);
+  factory EtherAmount.fromWei(dynamic amount) => EtherAmount.fromUnitAndValue(EtherUnit.wei, amount);
```
Comment (simolus3) [inline lib/src/core/amount.dart:71]: I would probably strip the `from` from these factory names. Each factory constructor should also some basic documentation, like
```dart
/// Constructs an amount of ether consisting of [amount] szabo, or [amount] trillion wei.
///
/// [amount] can either be a base10 [String], an [int] or a [BigInt].
factory EtherAmount.szabo(dynamic amount) => EtherAmount.fromUnitAndValue(EtherUnit.szabo, amount);
```

File: lib/src/core/client.dart [hunk 5]
```diff
@@ -233,7 +233,9 @@ class Web3Client {
   Future<TransactionInformation> getTransactionByHash(String transactionHash) {
     return _makeRPCCall<Map<String, dynamic>>(
             'eth_getTransactionByHash', [transactionHash])
-        .then((s) => TransactionInformation.fromMap(s));
+        .then((s) {
+            return TransactionInformation.fromMap(s);
+        } );
```
Comment (simolus3) [inline lib/src/core/client.dart:238]: Any reason for this change? I prefer the `=>` syntax for one-line functions.

File: lib/src/core/client.dart [hunk 6]
```diff
@@ -335,7 +339,7 @@ class Web3Client {
     EthereumAddress sender,
     EthereumAddress to,
     EtherAmount value,
-    BigInt amountOfGas,
+    int amountOfGas,/// gasLimit
```
Comment (simolus3) [inline lib/src/core/client.dart:342]: I think I'd prefer to keep a `BigInt` here, or why was this changed?

File: lib/src/core/client.dart [hunk 7]
```diff
@@ -346,10 +350,10 @@ class Web3Client {
         {
           if (sender != null) 'from': sender.hex,
           if (to != null) 'to': to.hex,
-          if (amountOfGas != null) 'gas': '0x${amountOfGas.toRadixString(16)}',
-          if (gasPrice != null)
-            'gasPrice': '0x${amountOfGas.toRadixString(16)}',
+          if (amountOfGas != null && amountOfGas != 0) 'gas': '0x${amountOfGas.toRadixString(16)}',
+          if (gasPrice != null) 'gasPrice': '0x${gasPrice.getInWei.toRadixString(16)}',
```
Comment (simolus3) [inline lib/src/core/client.dart:354]: Thanks for the `gasPrice` fix :+1:

File: lib/src/contracts/contract_invoker.dart [hunk 8]
```diff
@@ -0,0 +1,135 @@
...
+
+class ContractInvocation {
```
Comment (simolus3) [inline lib/src/contracts/contract_invoker.dart:16]: There should be documentation for this class and its members. Let me know if you need help.

File: lib/src/contracts/contract_invoker.dart [hunk 9]
```diff
@@ -0,0 +1,135 @@
...
+    List _params;
+
+    ContractInvocation(this.client, this.contract, this.function);
+
+    ContractInvocation parameters([dynamic args]) {
+
+        if ( args is Map ) {
+
+            for ( final p in function.parameters ) {
+                _params.add(args[p.name]);
```
Comment (simolus3) [inline lib/src/contracts/contract_invoker.dart:30]: We should probably throw an `ArgumentError` if `!args.containsKey(p.name)`

File: lib/src/contracts/contract_invoker.dart [hunk 10]
```diff
@@ -0,0 +1,135 @@
...
+        return client.callRaw(
+            sender: from,
+            contract: contract.address,
+            data: function.encodeCall(_params),
+            value: value,
+        ).then( (rawRsp) {
+
+            final r = function.decodeReturnValues(rawRsp);
+
+            if ( T is Map ) {
```
Comment (simolus3) [inline lib/src/contracts/contract_invoker.dart:58]: This will always be false - `T` is a `Type` and not a `Map`. To solve this, I'd split this into three methods:

- `callList<T>` returning a `List<T>`
- `callMap` returning a `Map<String, dynamic>`
- `call` returning `T` directly, but it should only work for functions only retuning a single value.

File: test/core/deployed_contract_test.dart [hunk 11]
```diff
@@ -0,0 +1,126 @@
...
+            ']',
+            /// contract name
+            'ERC20'
+        ),
+        contractAddress
+    );
+}
+
+
+void main() {
```
Comment (simolus3) [inline test/core/deployed_contract_test.dart:71]: We can't run this test in CI :( I wanted to setup something like ganache in the test runner so that we can run integration tests, but I never had the time. I think we need to remove this test if we don't have an rpc node to connect to.

File: lib/src/contracts/abi/abi.dart (hunk 1 above)
Comment (oblivioned) [reply lib/src/contracts/abi/abi.dart]: Well, I think so. So I changed it.

File: lib/src/contracts/abi/abi.dart (hunk 2 above)
Comment (oblivioned) [reply lib/src/contracts/abi/abi.dart]: The original "dart" can also be used in this way.

File: lib/src/contracts/contract_invoker.dart (hunk 3 above)
Comment (oblivioned) [reply lib/src/contracts/contract_invoker.dart]: I deleted them.

File: lib/src/core/client.dart (hunk 5 above)
Comment (oblivioned) [reply lib/src/core/client.dart]: I'm sorry, I forgot to delete it, and I won't adjust the code format at that time. Thank you for your guidance.

File: lib/src/core/client.dart (hunk 6 above)
Comment (oblivioned) [reply lib/src/core/client.dart]: At first I saw in "transaction. Callcontract" that gaslimit is the type of "int" used.
```dart
Transaction.callContract({
    @required DeployedContract contract,
    @required ContractFunction function,
    @required List<dynamic> parameters,
    this.from,
    this.maxGas,/// type is int
    this.gasPrice,
    this.value,
    this.nonce,
  })
```
Now that I've changed this back to the "bigint" type, and changed the field name to "gaslimit," I think it's easier to understand.

File: lib/src/contracts/contract_invoker.dart (hunk 9 above)
Comment (oblivioned) [reply lib/src/contracts/contract_invoker.dart]: I added this criterion

File: test/core/deployed_contract_test.dart (hunk 11 above)
Comment (oblivioned) [reply test/core/deployed_contract_test.dart]: I deleted this test

File: lib/src/contracts/contract_invoker.dart (hunk 10 above)
Comment (oblivioned) [reply lib/src/contracts/contract_invoker.dart]: The generic t in the call function, my intention is to define the type of data returned by the call result. In web3.js, for example:

```solidity
function f() external pure returns (uint a, uint b, uint c) {
     return (1, 2, 3);
}
```

The result will be
```
{
    [0] = BN (1).
    [1] = BN (2).
    [2] = BN (3).
    "A" = BN (1).
    "B" = BN (2).
    "C" = BN (3).
}
```

In this code, you can also use the

```dart
...call<List>()
/// [1,2,3]

call<Map>()
///  {"a":1, "b":2,"c":3}
```
So with this implementation, can't we use this syntax in "dart"?

**Comment by oblivioned** [comment]
Hello, simolus3. First of all, I think the implementation of "web3dart" is very good. In the process of reading the code, I learned a lot about dart, because the time to contact dart is very short and not very skilled,

You can see "dartfmt". Before I saw you, I still wondered how I should sort out the code format.

I'm a Chinese. My English level is poor. In the process of communication, I need to use translation software. Some expressions may not be able to express my meaning. Please forgive me. Here I specially speak Chinese and this passage to express my meaning most accurately.

As you can see, as far as my English level is concerned, the written documents can't describe the functions very well, so if you can, please help me.

Finally, thank you for your patience in reviewing the code and your comments on the changes. I have learned a lot of new knowledge.

There is also the latest support for "HDWallet". I hope it can be used. If my code needs to be modified, please help me. I'm just beginning to discuss these issues with you on GitHub. I'll cover a lot less.

Thank you!   


Original text:

        “simolus3” 你好，首先很我觉得"web3dart"的实现非常棒，在阅读代码的过程中，我学习到了很多关于dart的知识，因为接触dart这门语言的时间还很短，不是很熟练，

        你看就 "dartfmt"在看到你说之前，我还一直疑惑，我到底应该怎么样整理代码格式。

        我是一个中国人，凑巧我的英文水平很烂，交流的过程中，都需要通过翻译软件，有些表达可能词不达意，请原谅，这里我特意中中文和说这段话，是为了可以最准确的表达我的意思。

      关于前面说到补充文档的事情，如你所见，就我这个英文水平，写出来的文档并不能很好的描述功能，所以如果可以的话，请您帮忙。

最后，感谢您耐心的审查代码，和给我提出的修改意见，我掌握到了很多新的知识。

        这边还有最新提交的关于 HDWallet的相关支持，希望能派上用场，如果我的代码有需要修改的地方，也请劳烦您帮忙。对于在Github上和大家讨论这些问题，我也是刚刚开始，不到之处，多多包涵。

谢谢！

File: lib/src/contracts/contract_invoker.dart [hunk 12]
```diff
@@ -0,0 +1,118 @@
+part of 'package:web3dart/contracts.dart';
+
+class ContractInvocation {
+  final Web3Client client;
+  final DeployedContract contract;
+  final ContractFunction function;
```
Comment (simolus3) [inline lib/src/contracts/contract_invoker.dart:6]: I would make these parameters private:

```suggestion
  final Web3Client _client;
  final DeployedContract _contract;
  final ContractFunction _function;
```

File: lib/src/contracts/contract_invoker.dart [hunk 13]
```diff
@@ -0,0 +1,118 @@
...
+  List _params;
+
+  ContractInvocation(this.client, this.contract, this.function);
```
Comment (simolus3) [inline lib/src/contracts/contract_invoker.dart:9]: I think this constructor should be private as well, so that a `ContractInvocation` can only be created with `DeployedContract.invoker`.
```suggestion
  ContractInvocation._(this._client, this._contract, this._function);
```

File: lib/src/contracts/contract_invoker.dart [hunk 14]
```diff
@@ -0,0 +1,118 @@
...
+
+  ContractInvocation parameters([dynamic args]) {
+    if (args is Map) {
+      for (final p in function.parameters) {
+        if (!args.containsKey(p.name)) {
+          throw ArgumentError('Missing parameters "$p.name"');
+        }
+        _params.add(args[p.name]);
```
Comment (simolus3) [inline lib/src/contracts/contract_invoker.dart:17]: It seems like `_params` will always be null here?

File: lib/src/contracts/contract_invoker.dart [hunk 15]
```diff
@@ -0,0 +1,118 @@
...
+      }
+    } else if (args is List) {
+      _params = args;
+    } else if (args != null) {
+      _params = [args];
+    }
+
+    return this;
```
Comment (simolus3) [inline lib/src/contracts/contract_invoker.dart:25]: In Dart it's general not a good practice to return `this` for builder-like apis: https://dart.dev/guides/language/effective-dart/design#avoid-returning-this-from-methods-just-to-enable-a-fluent-interface

I would prefer to make `ContractInvocation` immutable:
- Add `_params` to the constructor: `ContractInvocation(this._client, this._contract, this._function, [this._params = const []]);`
- Make `_params` `final`.
- Create a new `params` list in this method and change `return this` to `return ContractInvocation(_client, _contract, _function, params)`

File: lib/src/contracts/contract_invoker.dart [hunk 16]
```diff
@@ -0,0 +1,118 @@
...
+class ContractInvocation {
+  final Web3Client client;
+  final DeployedContract contract;
+  final ContractFunction function;
+  List _params;
+
+  ContractInvocation(this.client, this.contract, this.function);
+
+  ContractInvocation parameters([dynamic args]) {
+    if (args is Map) {
```
Comment (simolus3) [inline lib/src/contracts/contract_invoker.dart:12]: This method overrides `_params`, which can be confusing when called twice. Let's add an
```suggestion
    assert(_params == null, 'Tried to call parameters multiple times');
    if (args is Map) {
```

File: lib/src/contracts/contract_invoker.dart [hunk 17]
```diff
@@ -0,0 +1,118 @@
...
+
+class ContractInvocation {
+  final Web3Client client;
+  final DeployedContract contract;
+  final ContractFunction function;
+  List _params;
+
+  ContractInvocation(this.client, this.contract, this.function);
+
+  ContractInvocation parameters([dynamic args]) {
```
Comment (simolus3) [inline lib/src/contracts/contract_invoker.dart:11]: I would suggest this documentation:
```suggestion
  /// Creates a new [ContractInvocation] instance that will call the function with provided [args].
  ///
  /// Args can either be:
  ///  - a `Map`, where the key is the parameter name and the value is the argument value
  //// - a `List`, which behaves like [ContractFunction.encodeCall].
  ////  -a simple value, which will be wrapped in a `List` and used as a single argument.
  ContractInvocation parameters([dynamic args]) {
```

File: lib/src/contracts/contract_invoker.dart [hunk 18]
```diff
@@ -0,0 +1,118 @@
...
+        return Future.value(ret as T);
+      } else if (T is List) {
+        return Future.value(function.decodeReturnValues(rawRsp) as T);
+      } else {
+        return Future.value(r.first as T);
+      }
+    });
+  }
+
+  Future<int> estimateGas({
```
Comment (simolus3) [inline lib/src/contracts/contract_invoker.dart:64]:
```suggestion
  /// Estimates the amount of gas that will be consumed when this call is sent.
  ///
  /// See also: [Web3Client.estimateGas].
  Future<int> estimateGas({
```

File: lib/src/contracts/contract_invoker.dart [hunk 19]
```diff
@@ -0,0 +1,118 @@
...
+    return client
+        .estimateGas(
+      sender: from,
+      to: contract.address,
+      value: value,
+      data: function.encodeCall(_params),
+    )
+        .then((gas) {
+      return Future.value(gas.toInt());
+    });
```
Comment (simolus3) [inline lib/src/contracts/contract_invoker.dart:77]: I would use an expression function here. We also don't need to wrap this in a `Future`:
```suggestion
        .then((gas) => gas.toInt());
```

File: lib/src/contracts/contract_invoker.dart [hunk 20]
```diff
@@ -0,0 +1,118 @@
...
+  }
+
+  Future<String> send({
```
Comment (simolus3) [inline lib/src/contracts/contract_invoker.dart:80]:
```suggestion
  /// Creates and sends a transaction to call this contract method with arguments from [parameters].
  /// See also: [Web3Client.sendTransaction].
  Future<String> send({
```

File: lib/src/contracts/contract_invoker.dart [hunk 21]
```diff
@@ -0,0 +1,118 @@
...
+    @required Credentials from,
+    EtherAmount gasPrice,
+    int gasLimit,
+    EtherAmount value,
+    int nonce = 0,
```
Comment (simolus3) [inline lib/src/contracts/contract_invoker.dart:85]: Let's just make `nonce` null by default. The `Web3Client` will fill the remaining data either way, which includes the `nonce` if null.

File: lib/src/contracts/contract_invoker.dart [hunk 22]
```diff
@@ -0,0 +1,118 @@
...
+      if (estimateGas > gasLimit) {
+        throw Exception(
+            'out of gas.[estimateGas:${estimateGas.toString()}, gasLimit:${gasLimit.toString()}]');
+      }
+    }
+
+    /// payload nonce
+    if (nonce == 0) {
+      nonce = await client.getTransactionCount(sender);
+    }
```
Comment (simolus3) [inline lib/src/contracts/contract_invoker.dart:101]: We can also remove this as the client takes care of it (in `_fillMissingData`).

File: lib/src/contracts/contract_invoker.dart (hunk 21 above)
Comment (simolus3) [inline lib/src/contracts/contract_invoker.dart:85]: Let's just make `nonce` null by default. The `Web3Client` will fill the remaining data either way, which includes the `nonce` if null.

File: lib/src/contracts/contract_invoker.dart (hunk 10 above)
Comment (simolus3) [reply lib/src/contracts/contract_invoker.dart]: We can use a similar syntax in Dart. However, we can't know whether `T` is a `List` or a `Map`.

I would still suggest to have three methods:

- `callList` returning a `List`
- `callMap` returning a `Map`
- `call` returning `dynamic`

I think it should be the users responsibility to cast the return value if that's desired.

Comment (oblivioned) [reply lib/src/contracts/contract_invoker.dart]: Indeed, I'm sure I understand your suggestion correctly now, but I suddenly feel that the addition of `call`,`callMap` or `callList` and call may make it difficult for users to determine the type, such as

```solidity
function returnVoid() external view; 
```

`callMap()` or `callList()` is used.

or

```solidity
function returnMultiple() external view returns (uint a, uint b);
```

`call()` is used.

Wait, input seems to provide convenience for users, but will it cause users to rely too much on the corresponding implementation and ignore the logic process of judging the return value?

Should we consider using callMap and callList instead of adding them

```dart
_function.decodeReturnValues(rspRaw);
```

The result of is directly returned as `call()`. The type is determined by the user.

Comment (simolus3) [reply lib/src/contracts/contract_invoker.dart]: Yeah, I think it might make sense to only have one `call` method returning `List<dynamic>` by using `decodeReturnValues`.

**Comment by oblivioned** [comment]
I think I need your help. I think I have developed my expected functions, but Travis CI has always been in the state of "the build failed" so that the code can't be merged. What should I do next? 

Thank you very much.

**Comment by simolus3** [comment]
You need to format `lib/src/core/client.dart` (just run `dartfmt -w lib/src/core/client.dart`).
